src/internal/repository-metadata-files/provider/artifacts.go

```diff
@@ -0,0 +1,147 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"registry-stable/internal/github"
+	"registry-stable/internal/provider"
+	"strings"
+	"sync"
+)
+
+type Artifact struct {
+	Name        string
+	DownloadURL string
+}
+
+type TargetArtifact struct {
+	Artifact
+
+	OS   string
+	Arch string
+}
+
+type VersionArtifacts struct {
+	ShaSumsArtifact          Artifact
+	ShaSumsSignatureArtifact Artifact
+	ManifestArtifact         Artifact
+	TargetArtifacts          []TargetArtifact
+}
+
+type VersionArtifactsMap map[string]VersionArtifacts
+
+type ShaSumResult struct {
+	Index     int
+	ShaSumMap map[string]string
+	Err       error
+}
+
+type ProtocolsResult struct {
+	Index     int
+	Protocols []string
+	Err       error
+}
+
+func getArtifacts(release github.GHRelease) VersionArtifacts {
+	var shaSumsArtifact Artifact
+	var shaSumsSignatureArtifact Artifact
+	var manifestArtifact Artifact
+	var targetArtifacts = make([]TargetArtifact, 0)
+
+	for _, asset := range release.ReleaseAssets.Nodes {
+		if platform := github.ExtractPlatformFromFilename(asset.Name); platform != nil {
+			targetArtifacts = append(targetArtifacts, TargetArtifact{
+				OS:   platform.OS,
+				Arch: platform.Arch,
+				Artifact: Artifact{
+					Name:        asset.Name,
+					DownloadURL: asset.DownloadURL,
+				},
+			})
+		} else if strings.HasSuffix(asset.Name, "SHA256SUMS") {
+			shaSumsArtifact = Artifact{
+				Name:        asset.Name,
+				DownloadURL: asset.DownloadURL,
+			}
+		} else if strings.HasSuffix(asset.Name, "SHA256SUMS.sig") {
+			shaSumsSignatureArtifact = Artifact{
+				Name:        asset.Name,
+				DownloadURL: asset.DownloadURL,
+			}
+		} else if strings.HasSuffix(asset.Name, "_manifest.json") {
+			manifestArtifact = Artifact{
+				Name:        asset.Name,
+				DownloadURL: asset.DownloadURL,
+			}
+		}
+	}
+
+	return VersionArtifacts{
+		ShaSumsArtifact:          shaSumsArtifact,
+		ShaSumsSignatureArtifact: shaSumsSignatureArtifact,
+		ManifestArtifact:         manifestArtifact,
+		TargetArtifacts:          targetArtifacts,
+	}
+}
+
+// enrichWithDataFromArtifacts performs all data enrichment necessary for the provider, that require artifact downloads
+// All necessary artifacts are downloaded in parallel, when calculating the SHAs and Protocols of each provider version
+func enrichWithDataFromArtifacts(ctx context.Context, versions []provider.Version, artifactsMap VersionArtifactsMap) ([]provider.Version, error) {
+	versionsCopy := versions
+
+	shaSumCh := make(chan ShaSumResult, len(versionsCopy))
+	protocolsCh := make(chan ProtocolsResult, len(versionsCopy))
+
+	var wg sync.WaitGroup
+	for i, v := range versionsCopy {
+		wg.Add(2)
+
+		go func(v provider.Version, i int) {
```

Comment:
I think this is fine for now, we probably want to make sure we limit the concurrency as to not hit rate limits as we ramp up.